### PR TITLE
Fix auto integer scaling when falling back to the `sharp` shader

### PR DESCRIPTION
**Pull request by johnnovak**
# Description

This fix/workaround will have to do for now. I will redesign the auto-shader switching in the future which will be a better fix (see my code comments for details, those are just some notes for myself).

There's also some general cleanup, and I've fixed up setting the minimum window size (see my comments for details).

## Related issues

- Fixes https://github.com/dosbox-staging/dosbox-staging/issues/4662

# Manual testing

### `sharp` shader fallback

- Repeated the test case in https://github.com/dosbox-staging/dosbox-staging/issues/4662

### Minimum window size

- You can also test it by resizing the window to its minimum size after startup, which will cause the `sharp` shader to be activated. The image will only snap to the full extents of the viewport if you resize the window a second time (just by a few pixels to make sure we're not triggering another shader switching).
- Tested the minimum window size is set as expected by resizing the window to the minimum on:
  - Windows 10, 3840x2160, 200% scaling
  - Windows 10, 3840x2160, 150% scaling
  - Windows 10, 1920x1080, 100% scaling
  - macOS, 3840x2160 (**"1920x1080 (Default)"** option in display settings)
  - macOS, 1920x1080 (**"1920x1080 (low resolution)"** option in display settings)

### Other `[sdl]` section settings

- Tested `dosbox --set "mute_when_inactive on" --set "pause_when_inactive on"` still works
- Tested `dosbox --set "keyboard_capture on"` still works and `keyboard_capture` can be toggled at runtime
- Tested `dosbox --set "window_transparency 50"` still works
- Tested `dosbox --set "window_decorations off"` still works and `window_decorations` can be set at runtime
 
The change has been manually tested on:

- [x] Windows
- [x] macOS
- [x] Linux

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] my change has been manually tested on Windows, macOS, and Linux.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/gui/sdl_gui.cpp [hunk 1]
```diff
@@ -591,51 +591,27 @@ static bool is_aspect_ratio_correction_enabled()
...
-	//          FallbackWindowSize.y);
-
-	// Keep the SDL minimum allowed window size in lock-step with the
-	// fallback dimensions. If these aren't linked, the window can obscure
-	// the content (or vice-versa).
-	if (!sdl.window) {
-		return;
-		// LOG_WARNING("SDL: Tried setting window minimum size,"
-		//             " but the SDL window is not available yet");
-	}
```
Comment (weirddan455) [inline src/gui/sdl_gui.cpp:621]: Should we be getting rid of this nullptr check? Feel like this could blow up in our face :laughing:

**Review by weirddan455 — APPROVED** [review]
Looks good. I tested that it fixed the issue I reported. Just one small comment.

File: src/gui/sdl_gui.cpp (hunk 1 above)
Comment (johnnovak) [reply src/gui/sdl_gui.cpp:621]: I think the check was a bit over the top. I'm happy to put an `assert(sdl.window)` at the top of the function, but look all the occurences of using nullable pointers in this file.

This is called once after creating the window. Calling this before the window exists is thus a programming error because there is no dynamic runtime behaviour involved which could warrant conditional logic and perhaphs debug logging.

Comment (weirddan455) [reply src/gui/sdl_gui.cpp:621]: Ok, just checking. You can put an assert there if you want or just merge when ready.

**Comment by johnnovak** [comment]
@weirddan455 You know what, adding asserts for `sdl.window` and `sdl.renderer` to every usage site is just a good idea, so I've done it.

Quit a bit of useless and redundant code fell out as a result of this exercise, and I've had to rearrange the order of things a little bit because some of that was just wrong. Some of the new assertions got tripped because it turns out many (or all?) SDL calls silently ignore null `SDL_Window*` arguments. Yeah... I don't like relying on that. 😐 

I've managed to do some further simple cleanups as well after the init order became clearer and more explicit. One could keep polishing `sdl_gui.cpp` to ages; I prefer to do it in little chunks quickly when the inspiration strikes like now😄  Lots of small commits, but should be very easy to review.

I've extended the testing section with the extra tests I've done, and finally I made sure every commit compiles & Staging starts up by executing the following for every commit:

```bash
cmake --build --preset=debug-macos && \
  ./build/debug-macos/Debug/dosbox && \
  ./build/debug-macos/Debug/dosbox --set "output texture" && \
  git rebase --continue
```
